gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/reactive/handlers/api/security/SecurityChain.java

```diff
@@ -0,0 +1,122 @@
+import io.gravitee.gateway.reactive.handlers.api.security.handler.SecurityHandler;
+import io.gravitee.gateway.reactive.handlers.api.security.handler.SecurityHandlerFactory;
+import io.gravitee.gateway.reactive.policy.PolicyManager;
+import io.reactivex.Completable;
+import io.reactivex.Flowable;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link SecurityChain} is a special chain dedicated to execute policy associated with plans.
+ * The security chain is responsible to create {@link SecurityHandler} for each plan of the api and executed them in order.
+ * Only the first {@link SecurityHandler} that can handle the current request is executed.
+ * The result of the security chain execution depends on this {@link SecurityHandler} execution.
+ *
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class SecurityChain {
+
+    private static final Logger log = LoggerFactory.getLogger(SecurityChain.class);
+    public static final String SKIP_SECURITY_CHAIN = "skip-security-chain";
+    protected static final String PLAN_UNRESOLVABLE = "GATEWAY_PLAN_UNRESOLVABLE";
+    protected static final String ATTR_SECURITY_HANDLED = "security-handled";
+    protected static final String UNAUTHORIZED_MESSAGE = "Unauthorized";
+
+    private final Flowable<SecurityHandler> chain;
+
+    public SecurityChain(Api api, PolicyManager policyManager) {
+        chain =
+            Flowable.fromIterable(
+                api
+                    .getPlans()
+                    .stream()
+                    .map(plan -> SecurityHandlerFactory.forPlan(plan, policyManager))
+                    .filter(Objects::nonNull)
+                    .sorted(Comparator.comparingInt(SecurityHandler::order))
+                    .collect(Collectors.toList())
+            );
+    }
+
+    /**
+     * Executes the security chain by executing all the {@link SecurityHandler}s in an ordered sequence.
+     * It's up to each {@link SecurityHandler} to provider its order. The lower is the order, the highest priority is.
+     * The result of the security chain execution depends on the first {@link SecurityHandler} able to handle the request.
+     * If no {@link SecurityHandler} has been executed because there is no {@link SecurityHandler} in the chain or none of them can handle the request,
+     * then the chain is interrupted with a 401 response status and the {@link Completable} returns an error.
+     *
+     * @param ctx the current execution context.
+     * @return a {@link Completable} that completes if the request has been successfully handled by a {@link SecurityHandler} or returns
+     * an error if no {@link SecurityHandler} can handle the request or the {@link SecurityHandler} failed.
+     */
+    public Completable execute(RequestExecutionContext ctx) {
```

Comment:
```suggestion
public Completable execute(RequestExecutionContext ctx) {
        return defer(() -> {
            if (!Objects.equals(true, ctx.getAttribute(SKIP_SECURITY_CHAIN))) {
                return chain
                    .flatMapSingle(policy -> continueChain(ctx, policy), false, 1)
                    .any(securityHandled -> securityHandled)
                    .flatMapCompletable(securityHandled -> {
                        if (!securityHandled) {
                            return ctx.interruptWith(new ExecutionFailure(UNAUTHORIZED_401, PLAN_UNRESOLVABLE, UNAUTHORIZED_MESSAGE));
                        }
                        return Completable.complete();
                    })
                    .doOnSubscribe(disposable -> log.debug("Executing security chain"));
            }

            log.debug("Skipping security chain because it has been explicitly required");
            return Completable.complete();
        });
    }
    }```

Reply:
change has been directly commited

Reply:
Looks definitely better 👍